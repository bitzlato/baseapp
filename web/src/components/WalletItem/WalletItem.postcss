:root {
  --wallets-wallet-item-color: rgb(102 102 102);
  --wallets-wallet-item-color-hover: rgb(255 255 255);
  --wallets-wallet-item-bg-hover: rgb(64 69 76);
  --wallets-wallet-item-after-bg-active: rgb(var(--rgb-brand-blue));
}

:global(.light-mode) {
  --wallets-wallet-item-color: rgb(153 153 153);
  --wallets-wallet-item-color-hover: rgb(0 0 0);
  --wallets-wallet-item-bg-hover: rgb(240 243 249);
  --wallets-wallet-item-after-bg-active: rgb(91 142 229);
}

.item {
  background: transparent;
  border: 0;
  border-bottom: 1px solid var(--wallets-item-border-color);
  color: var(--wallets-wallet-item-color);
  min-width: 320px;
  position: relative;

  &:hover,
  &.itemActive {
    background-color: var(--wallets-wallet-item-bg-hover);
    color: var(--wallets-wallet-item-color-hover);
  }

  &.itemActive {
    border-bottom-color: transparent;

    &::after {
      background-color: var(--wallets-wallet-item-after-bg-active);
      content: "";
      height: 100%;
      left: 0;
      position: absolute;
      width: 3px;
    }
  }
}

.itemHeader {
  border-bottom: 1px solid var(--wallets-item-border-color);
  padding-bottom: var(--gap);
  padding-top: calc(var(--gap) * 2);

  @media (--breakpoint-down-md) {
    padding-left: calc(var(--gap) * 3 + 36px + var(--gap) * 2);
    padding-right: calc(var(--gap) * 3);
  }

  @media (--breakpoint-up-md) {
    padding-left: calc(var(--gap) * 2 + 36px + var(--gap) * 2);
    padding-right: calc(var(--gap) * 2);
  }
}
