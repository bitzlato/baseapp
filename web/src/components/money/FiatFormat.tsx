import { Currency, FormatOptions } from '@bitzlato/money-js';
import { FC } from 'react';
import { AmountFormat, AmountFormatProps } from 'web/src/components/AmountFormat/AmountFormat';

const CRYPTO_CURRENCY_CODES_WITH_CENTS = new Set(['MCR', 'USDT', 'USDC', 'DOGE', 'DAI', 'MDT']);

const getOptionsByCode = (code: string): FormatOptions => {
  return CRYPTO_CURRENCY_CODES_WITH_CENTS.has(code)
    ? {
        minFractionDigits: 0,
        maxFractionDigits: 2,
      }
    : {
        minFractionDigits: 0,
        maxFractionDigits: 0,
      };
};

export interface Props extends AmountFormatProps {
  cryptoCurrency: Currency;
}

export const FiatFormat: FC<Props> = ({ cryptoCurrency, ...props }) => (
  <AmountFormat {...props} {...getOptionsByCode(cryptoCurrency.code)} />
);
