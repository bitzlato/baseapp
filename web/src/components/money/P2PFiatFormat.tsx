import { Currency } from '@bitzlato/money-js';
import { FC } from 'react';
import { AmountFormat, AmountFormatProps } from 'web/src/components/AmountFormat/AmountFormat';
import { getP2PFiatOptionsByCode } from '../AmountFormat/getFormatOptionsByLanguage';

export interface Props extends AmountFormatProps {
  cryptoCurrency: Currency;
}

export const P2PFiatFormat: FC<Props> = ({ cryptoCurrency, ...props }) => (
  <AmountFormat {...props} {...getP2PFiatOptionsByCode(cryptoCurrency.code)} />
);
