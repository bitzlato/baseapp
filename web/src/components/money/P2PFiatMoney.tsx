import { Currency } from '@bitzlato/money-js';
import { FC } from 'react';
import { getP2PFiatOptionsByCode } from '../AmountFormat/getFormatOptionsByLanguage';
import { MoneyFormat, MoneyFormatProps } from '../MoneyFormat/MoneyFormat';

export interface Props extends MoneyFormatProps {
  cryptoCurrency: Currency;
}

export const P2PMoneyFormat: FC<Props> = ({ cryptoCurrency, ...props }) => (
  <MoneyFormat {...props} {...getP2PFiatOptionsByCode(cryptoCurrency.code)} />
);
