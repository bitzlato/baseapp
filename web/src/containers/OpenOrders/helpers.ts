export const getTriggerSign = (ordType: string, ordSide: string) => {
  if (ordType.includes('stop')) {
    if (ordSide === 'buy') {
      return '≥';
    }
    return '≤';
  }

  if (ordType.includes('take')) {
    if (ordSide === 'buy') {
      return '≤';
    }
    return '≥';
  }
};
