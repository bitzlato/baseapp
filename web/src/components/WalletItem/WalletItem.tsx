import { FC } from 'react';
import cn from 'classnames';
import { Money } from '@bitzlato/money-js';
import { getCurrencyCodeSymbol } from 'src/helpers/getCurrencySymbol';
import { CryptoCurrencyIcon } from 'src/components/CryptoCurrencyIcon/CryptoCurrencyIcon';
import { AmountFormat } from 'src/components/AmountFormat/AmountFormat';
import { Box } from 'src/components/Box/Box';
import { MoneyFormat } from 'src/components/MoneyFormat/MoneyFormat';
import s from './WalletItem.postcss';

export interface WalletItemData {
  name: string;
  currency: string;
  balance: Money;
  balanceP2P: Money | undefined;
  balanceMarket: Money | undefined;
  approximate: Money;
  locked: Money;
  icon: string;
  hasTransfer: boolean;
  index: number;
}

interface Props {
  active?: boolean | undefined;
  wallet: WalletItemData;
  onClick?: (() => void) | undefined;
  isMobileDevice?: boolean | undefined;
}

export const WalletItem: FC<Props> = ({ wallet, active, onClick, isMobileDevice }) => {
  const color = isMobileDevice ? 'primary' : undefined;
  return (
    <Box
      row
      spacing="2"
      padding={isMobileDevice ? '2X3' : '2'}
      as="button"
      className={cn(s.item, active && s.itemActive)}
      type="button"
      onClick={onClick}
    >
      <CryptoCurrencyIcon size="medium" currency={wallet.currency} />
      <Box grow row align="start" justify="between">
        <Box col align="start" textAlign="start">
          <Box textColor={color} className={s.title}>
            {getCurrencyCodeSymbol(wallet.currency)}
          </Box>
          <span className={s.description}>{wallet.name}</span>
        </Box>
        <Box col align="end">
          <Box className={s.title} textColor={color}>
            <AmountFormat money={wallet.balance} />
          </Box>
          {wallet.approximate.isZero() ? null : (
            <Box className={s.description} textColor={color}>
              <Box as="span" textColor="secondary">
                &asymp;
              </Box>
              &nbsp;
              <MoneyFormat money={wallet.approximate} />
            </Box>
          )}
        </Box>
      </Box>
    </Box>
  );
};
