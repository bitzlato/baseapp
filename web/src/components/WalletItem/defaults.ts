import { ZERO_PENCE } from 'web/src/helpers/money';
import { WalletItemData } from './WalletItem';

export const DEFAULT_WALLET_ITEM: WalletItemData = {
  name: '',
  currency: '',
  balanceTotal: ZERO_PENCE,
  balanceP2P: ZERO_PENCE,
  balanceMarket: ZERO_PENCE,
  approximate: ZERO_PENCE,
  locked: ZERO_PENCE,
  hasDepositWithdraw: false,
  hasTransfer: false,
  hasGift: false,
  index: -1,
};
