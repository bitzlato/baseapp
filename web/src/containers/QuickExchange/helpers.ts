import type { Currency } from '@bitzlato/money-js';
import { lowerBound } from 'src/helpers/lowerBound';
import type { Market, Wallet } from '../../modules';

export interface DropdownItem {
  code: string;
  match: boolean;
}

interface SwapState {
  fromList: DropdownItem[];
  toList: DropdownItem[];
  market?: Market;
}

export function getCurrencies(markets: Market[], base: string, quote: string): SwapState {
  let market: Market | undefined;
  const fromList: DropdownItem[] = [];
  const toList: DropdownItem[] = [];

  for (let m of markets) {
    const sell = m.base_unit === base && m.quote_unit === quote;
    if (sell || (m.base_unit === quote && m.quote_unit === base)) {
      market = m;
    }

    if (quote !== m.base_unit) insertUnique(fromList, m.base_unit, m.quote_unit === quote);
    if (quote !== m.quote_unit) insertUnique(fromList, m.quote_unit, m.base_unit === quote);

    if (base !== m.base_unit) insertUnique(toList, m.base_unit, m.quote_unit === base);
    if (base !== m.quote_unit) insertUnique(toList, m.quote_unit, m.base_unit === base);
  }

  fromList.sort(compareItems);
  toList.sort(compareItems);

  return {
    fromList,
    toList,
    market,
  };
}

function insertUnique(items: DropdownItem[], code: string, match: boolean) {
  const value = { code, match };
  const index = lowerBound(items, value, lessName);
  const found = index !== items.length && !lessName(value, items[index]);
  if (!found) {
    items.splice(index, 0, value);
  } else {
    const d = items[index];
    d.match ||= match;
  }
}

function lessName(a: DropdownItem, b: DropdownItem) {
  return a.code < b.code;
}

function compareItems(a: DropdownItem, b: DropdownItem): number {
  if (a.match !== b.match) {
    return Number(b.match) - Number(a.match);
  }
  return a.code === b.code ? 0 : a.code < b.code ? -1 : 1;
}

export function getWallet(currency: string, wallets: Wallet[]): Wallet | undefined {
  const lowerCurrency = currency.toLowerCase();
  return wallets.find((w) => w.currency.code.toLowerCase() === lowerCurrency);
}

export function getCurrency(code: Currency['code'], minorUnit: Currency['minorUnit']) {
  return {
    code,
    minorUnit,
  };
}

export function getItem(code: string, match: boolean): DropdownItem {
  return { code, match };
}
