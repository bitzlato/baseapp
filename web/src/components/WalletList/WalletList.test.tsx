import { mount, shallow } from 'enzyme';
import { BTC_CCY, createCcy, createMoney, USD_CCY } from 'src/helpers/money';
import { TestComponentWrapper } from 'src/lib/test/wrapper';
import { WalletList, WalletListProps } from '..';
import { WalletItemData } from '../WalletItem/WalletItem';

const onWalletSelectionChange = jest.fn();

const ETH_CCY = createCcy('ETH', 8);

const walletItems: WalletItemData[] = [
  {
    name: 'Bitcoin',
    currency: 'BTC',
    balanceTotal: createMoney('1', BTC_CCY),
    balanceP2P: createMoney('0.5', BTC_CCY),
    balanceMarket: createMoney('0.5', BTC_CCY),
    approximate: createMoney('41000', USD_CCY),
    locked: createMoney('1', BTC_CCY),
    hasDepositWithdraw: true,
    hasTransfer: true,
    hasGift: true,
    index: 0,
  },
  {
    name: 'Ethereum',
    currency: 'ETH',
    balanceTotal: createMoney('1', ETH_CCY),
    balanceP2P: createMoney('0.5', ETH_CCY),
    balanceMarket: createMoney('0.5', ETH_CCY),
    approximate: createMoney('41000', USD_CCY),
    locked: createMoney('1', ETH_CCY),
    hasDepositWithdraw: true,
    hasTransfer: true,
    hasGift: true,
    index: 1,
  },
];

const defaultProps: WalletListProps = {
  activeIndex: 0,
  onWalletSelectionChange,
  walletItems,
};

const setup = (props: Partial<WalletListProps> = {}) =>
  shallow(
    <TestComponentWrapper>
      <WalletList {...{ ...defaultProps, ...props }} />
    </TestComponentWrapper>,
  );

const setup2 = (props: Partial<WalletListProps> = {}) =>
  mount(
    <TestComponentWrapper>
      <WalletList {...{ ...defaultProps, ...props }} />
    </TestComponentWrapper>,
  );

describe('WalletList', () => {
  it('should render', () => {
    expect(setup().render()).toMatchSnapshot();
  });

  it('should handle onWalletSelectionChange callback when an element is pressed', () => {
    const wrapper = setup2();
    const first = wrapper.find('button').first();
    first.simulate('click');
    expect(onWalletSelectionChange).toHaveBeenCalled();
    expect(onWalletSelectionChange).toHaveBeenCalledTimes(1);
  });
});
