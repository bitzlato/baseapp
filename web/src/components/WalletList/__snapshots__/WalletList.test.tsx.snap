// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletList should render 1`] = `
<div
  className="cr-wallet-list"
>
  <WalletItem
    active={true}
    key="0"
    onClick={[Function]}
    wallet={
      Object {
        "active": false,
        "balance": "456",
        "currency": "BTC",
        "explorerAddress": "https://testnet.blockchain.info/address/#{address}",
        "explorerTransaction": "https://testnet.blockchain.info/tx/#{txid}",
        "fee": 0.123,
        "fixed": 8,
        "icon_id": "BTC",
        "locked": "1",
        "name": "Bitcoin",
        "type": "fiat",
      }
    }
  />
  <WalletItem
    active={false}
    key="1"
    onClick={[Function]}
    wallet={
      Object {
        "active": false,
        "balance": "456",
        "currency": "USD",
        "fee": 0.123,
        "fixed": 8,
        "icon_id": "USD",
        "locked": "100",
        "name": "United states Dollar",
        "type": "coin",
      }
    }
  />
  <WalletItem
    active={false}
    key="2"
    onClick={[Function]}
    wallet={
      Object {
        "active": false,
        "balance": "2",
        "currency": "BTC",
        "fee": 0.3,
        "fixed": 8,
        "icon_id": "BTC",
        "locked": "0.4",
        "name": "Bitcoin - 2",
        "type": "fiat",
      }
    }
  />
</div>
`;
